src/XR/features/WebXRNearInteraction.ts

```diff
@@ -321,148 +361,172 @@ export class WebXRNearInteraction extends WebXRAbstractFeature {
+
                 let nearInteractionAtOrigin = false;
                 let nearInteraction = nearInteractionInfo && nearInteractionInfo.pickedPoint && nearInteractionInfo.hit;
                 if (nearInteractionInfo?.pickedPoint) {
                     nearInteractionAtOrigin = nearInteractionInfo.pickedPoint.x === 0 && nearInteractionInfo.pickedPoint.y === 0 && nearInteractionInfo.pickedPoint.z === 0;
                 }
                 if (nearInteraction && !nearInteractionAtOrigin) {
-                    result = nearInteractionInfo;
-                    controllerData.nearInteraction = true;
+                    result = nearInteractionInfo!;
                 }
                 return result;
             };
 
             let pick = null;
-            let hoverRange = false;
+
             // near interaction hover
             if (controllerData.hoverIndexMeshTip) {
                 let utilitySceneHoverPick = null;
-                let utilitySceneHoverGrab = null;
                 if (this._options.useUtilityLayer && this._utilityLayerScene) {
-                    utilitySceneHoverPick = this._pickWithMesh(controllerData.hoverIndexMeshTip, this._utilityLayerScene, false, this.nearPickPredicate);
+                    utilitySceneHoverPick = this._pickWithSphere(controllerData, this._hoverRadius, this._utilityLayerScene, (mesh: AbstractMesh) =>
+                        this._nearInteractionPredicate(mesh)
+                    );
                 }
-                let originalSceneHoverPick = this._pickWithMesh(controllerData.hoverIndexMeshTip, this._scene, false, this.nearPickPredicate);
-                if (this._options.useUtilityLayer && this._utilityLayerScene) {
-                    utilitySceneHoverGrab = this._pickWithMesh(controllerData.hoverIndexMeshTip, this._utilityLayerScene, false, this.nearGrabPredicate);
-                }
-                let originalSceneHoverGrab = this._pickWithMesh(controllerData.hoverIndexMeshTip, this._scene, false, this.nearGrabPredicate);
+                let originalSceneHoverPick = this._pickWithSphere(controllerData, this._hoverRadius, this._scene, (mesh: AbstractMesh) => this._nearInteractionPredicate(mesh));
 
                 let hoverPickInfo = accuratePickInfo(originalSceneHoverPick, utilitySceneHoverPick);
-                let hoverGrabInfo = accuratePickInfo(originalSceneHoverGrab, utilitySceneHoverGrab);
-                if ((hoverPickInfo && hoverPickInfo.hit) || (hoverGrabInfo && hoverGrabInfo.hit)) {
-                    hoverRange = true;
+                if (hoverPickInfo && hoverPickInfo.hit) {
                     // turn off far interaction if in the hover range
                     if (this._farInteractionFeature) {
                         this._farInteractionFeature.detach();
                     }
                     pick = populateNearInteractionInfo(hoverPickInfo);
+                    if (pick.hit) {
+                        controllerData.hoverInteraction = true;

```

Comment:
We probably shouldn't disable the far interaction feature if populateNearInteractionInfo gives an invalid pick info, since that would cause a feature detach/attach every frame where it happens.

Actually, does it make more sense to place both the detach and attach logic together down below, as an if/else conditional?

Reply:
yep, good catch